mpi_test.py stopped working.

To reproduce, 

```
$ mpirun -np 2 pytest tests/python/mpi_test.py

_____________________________________________________________________________________ ERROR collecting tests/python/mpi_test.py ______________________________________________________________________________________
tests/python/mpi_test.py:43: in <module>
    o = fn.execute(inputs)
nvfuser/__init__.py:215: in execute
    self._setup_schedule(inputs)
E   RuntimeError:  INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/fusion_executor/executor_utils.cpp":731, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. When trying to run the provided host program, there was an error with the provided input 0. Provided input was:
E     Tensor(sizes=[1, 4], stride=[4, 1], dtype=float, device=cuda:0, data_ptr=0x7d88c0e00000)
E     Fusion input is:
E     T0_g_float[ iS0{2}, iS1{4} ]
E     Expr eval provided the error:
E   """Tried to bind to a constant value: 2 as 1"""
E
E   Exception raised from bindInputs at /opt/pytorch/nvfuser/csrc/fusion_executor/executor_utils.cpp:731 (most recent call first):
E   frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x7d88d979b17f in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #1: nvfuser::nvfErrorFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x53 (0x7d88d9b00223 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #2: <unknown function> + 0x13ce26 (0x7d88d97a6e26 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #3: <unknown function> + 0x45668c (0x7d88d9ac068c in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #4: nvfuser::python_frontend::FusionDefinition::setupSchedule(c10::ArrayRef<c10::IValue> const&) + 0x186 (0x7d88d9f3ea36 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #5: <unknown function> + 0x1b367d (0x7d88d981d67d in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #6: <unknown function> + 0x22e603 (0x7d88d9898603 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #7: <unknown function> + 0x2bb660 (0x7d88d9925660 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
E   frame #8: <unknown function> + 0x15adae (0x64c764b7bdae in /usr/bin/python3)
E   frame #9: _PyObject_MakeTpCall + 0x25b (0x64c764b7252b in /usr/bin/python3)
E   frame #10: <unknown function> + 0x16952b (0x64c764b8a52b in /usr/bin/python3)
E   frame #11: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #12: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #13: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #14: <unknown function> + 0x140096 (0x64c764b61096 in /usr/bin/python3)
E   frame #15: PyEval_EvalCode + 0x86 (0x64c764c56f66 in /usr/bin/python3)
E   frame #16: <unknown function> + 0x23ba2d (0x64c764c5ca2d in /usr/bin/python3)
E   frame #17: <unknown function> + 0x15b909 (0x64c764b7c909 in /usr/bin/python3)
E   frame #18: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #19: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #20: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #21: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #22: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #23: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #24: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #25: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #26: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #27: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #28: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #29: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #30: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #31: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #32: _PyEval_EvalFrameDefault + 0x19b6 (0x64c764b65c16 in /usr/bin/python3)
E   frame #33: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #34: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #35: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #36: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #37: <unknown function> + 0x166d4e (0x64c764b87d4e in /usr/bin/python3)
E   frame #38: _PyObject_GenericGetAttrWithDict + 0x501 (0x64c764b7a4a1 in /usr/bin/python3)
E   frame #39: PyObject_GetAttr + 0x4d (0x64c764b7895d in /usr/bin/python3)
E   frame #40: _PyEval_EvalFrameDefault + 0x61c5 (0x64c764b6a425 in /usr/bin/python3)
E   frame #41: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #42: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #43: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #44: _PyEval_EvalFrameDefault + 0x8cb (0x64c764b64b2b in /usr/bin/python3)
E   frame #45: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #46: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #47: <unknown function> + 0x169251 (0x64c764b8a251 in /usr/bin/python3)
E   frame #48: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #49: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #50: _PyEval_EvalFrameDefault + 0x2a49 (0x64c764b66ca9 in /usr/bin/python3)
E   frame #51: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #52: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #53: <unknown function> + 0x169251 (0x64c764b8a251 in /usr/bin/python3)
E   frame #54: _PyEval_EvalFrameDefault + 0x64e2 (0x64c764b6a742 in /usr/bin/python3)
E   frame #55: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #56: _PyObject_FastCallDictTstate + 0x16d (0x64c764b7176d in /usr/bin/python3)
E   frame #57: _PyObject_Call_Prepend + 0x5c (0x64c764b875ac in /usr/bin/python3)
E   frame #58: <unknown function> + 0x281370 (0x64c764ca2370 in /usr/bin/python3)
E   frame #59: _PyObject_MakeTpCall + 0x25b (0x64c764b7252b in /usr/bin/python3)
E   frame #60: _PyEval_EvalFrameDefault + 0x7611 (0x64c764b6b871 in /usr/bin/python3)
E   frame #61: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
E   frame #62: _PyEval_EvalFrameDefault + 0x6d5 (0x64c764b64935 in /usr/bin/python3)
E   frame #63: _PyFunction_Vectorcall + 0x7c (0x64c764b7c6ac in /usr/bin/python3)
============================================================================================== short test summary info ===============================================================================================
ERROR tests/python/mpi_test.py - RuntimeError:  INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/fusion_executor/executor_utils.cpp":731, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. When try...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================== 1 error in 1.31s ==================================================================================================
```

I guess it started with https://github.com/NVIDIA/Fuser/pull/2635. However, I believe that's a right fix and shouldn't be rolled back. 

Instead, the root cause I believe is that bindInputs (called by concretization, which is in turn called by `FusionDefinition._setup_schedule`) happens before `FusionDefinition.schedule`: https://github.com/NVIDIA/Fuser/blob/7090a639ca9aa5154ef11c3cdf34374a1a4fc357/nvfuser/__init__.py#L215-L216. Without knowing the input is parallelized on DIDx, concretization incorrectly attempts to bind the device dimension extent with 1.

--- a/csrc/python_frontend/fusion_definition.cpp
+++ b/csrc/python_frontend/fusion_definition.cpp
@@ -9,7 +9,6 @@
 #include <fusion_executor/executor_kernel_arg.h>
 #include <fusion_profiler.h>
 #include <instrumentation.h>
-#include <multidevice/communicator.h>
 #include <options.h>
 #include <python_frontend/fusion_cache.h>
 #include <python_frontend/fusion_definition.h>
@@ -241,15 +240,6 @@ void FusionDefinition::setupSchedule(const at::ArrayRef<c10::IValue>& inputs) {
 void FusionDefinition::finalizeSchedule(
     const at::ArrayRef<c10::IValue>& inputs) {
   FUSER_PERF_SCOPE("FusionDefinition::finalizeSchedule");
-  // TODO: remove when multidevice executor integration is done natively
-  Fusion* fusion = user_sched_->schedule.get();
-  std::vector<TensorView*> tvs = fusion->allTvs();
-  if (std::any_of(tvs.begin(), tvs.end(), [](Val* v) {
-        return v->isA<TensorView>() && v->as<TensorView>()->hasDeviceMesh();
-      })) {
-    multidevice_executor_ = std::make_unique<MultiDeviceExecutor>(
-        std::make_unique<Fusion>(*fusion), Communicator::getInstance());
-  }
 
   FusionGuard::setCurFusion(prev_fusion_);
   user_sched_->runtime_info.reset();
@@ -291,10 +281,6 @@ std::vector<at::Tensor> FusionDefinition::execute(
   NVF_CHECK(id().has_value(), "Valid fusion schedule is not available!");
 
   auto scheds = fusionCache()->queryFusionSchedules(id().value());
-
-  if (multidevice_executor_) {
-    return multidevice_executor_->runWithInput(inputs.vec());
-  }
 
   std::vector<at::Tensor> outputs;
   if (profile) {

--- a/csrc/python_frontend/fusion_definition.h
+++ b/csrc/python_frontend/fusion_definition.h
@@ -10,7 +10,6 @@
 #include <iostream>
 
 #include <kernel_cache.h>
-#include <multidevice/executor.h>
 #include <python_frontend/fusion_state.h>
 #include <visibility.h>
 
@@ -317,12 +316,6 @@ class NVF_API FusionDefinition : public FusionState {
 
  private:
   mutable std::optional<std::string> debug_output_ = std::nullopt;
-
-  //! The reason we have these is due to the lack of cache for multidevice
-  //! executor. The assumption is that the same multidevice_executor can handle
-  //! device switches. This should be removed after multidevice executor is
-  //! properly integrated in the runtime.
-  mutable std::unique_ptr<MultiDeviceExecutor> multidevice_executor_;
 };
 
 } // namespace nvfuser::python_frontend

--- a/nvfuser/__init__.py
+++ b/nvfuser/__init__.py
@@ -133,6 +133,11 @@ class FusionDefinition(_C._FusionDefinition):
     def definition(self):
         raise NotImplementedError("definition() should be implemented by child class!")
 
+    def multidevice_schedule(self):
+        raise NotImplementedError(
+            "multdevice_schedule() should be implemented by child class!"
+        )
+
     def schedule(self):
         raise NotImplementedError("schedule() should be implemented by child class!")
 
@@ -201,6 +206,9 @@ class FusionDefinition(_C._FusionDefinition):
             self._setup_definition()
             self.definition()
             self._finalize_definition()
+
+        if type(self).multidevice_schedule != FusionDefinition.multidevice_schedule:
+            self.multidevice_schedule()
 
         # If schedule is defined by child class and schedule is not defined for
         # inputs, make a schedule.

--- a/nvfuser/__init__.py
+++ b/nvfuser/__init__.py
@@ -213,11 +213,19 @@ class FusionDefinition(_C._FusionDefinition):
             self.definition()
             self._finalize_definition()
 
-        self.multidevice_schedule()
+        defined_multidevice_schedule = (
+            type(self).multidevice_schedule != FusionDefinition.multidevice_schedule
+        )
+        defined_schedule = type(self).schedule != FusionDefinition.schedule
+        assert not (
+            defined_multidevice_schedule and defined_schedule
+        ), "I haven't tested what if both are defined. We don't plan to support this use case although it may just work."
+
+        if defined_multidevice_schedule:
+            self.multidevice_schedule()
 
         # If schedule is defined by child class and schedule is not defined for
         # inputs, make a schedule.
-        defined_schedule = type(self).schedule != FusionDefinition.schedule
         if defined_schedule and not self._exist_schedule(inputs):
             self._setup_schedule(inputs)
             self.schedule()

--- a/nvfuser/__init__.py
+++ b/nvfuser/__init__.py
@@ -133,10 +133,16 @@ class FusionDefinition(_C._FusionDefinition):
     def definition(self):
         raise NotImplementedError("definition() should be implemented by child class!")
 
+    # Unlike `schedule`, `multidevice_schedule` is designed for inter-device
+    # scheduling, The scheduling is done before concretization and therefore
+    # before pre-segmentation. `schedule` however assumes the FusionDefinition
+    # has been concretized and pre-segmented, and therefore requires
+    # `_setup_schedule` and `_finalize_schedule` to be called before and after.
+    #
+    # Note: there's a plan to embed multidevice schedules into FusionDefinition
+    # as annotating nodes. This may eventually replace `multidevice_schedule`.
     def multidevice_schedule(self):
-        raise NotImplementedError(
-            "multdevice_schedule() should be implemented by child class!"
-        )
+        pass
 
     def schedule(self):
         raise NotImplementedError("schedule() should be implemented by child class!")
@@ -207,8 +213,7 @@ class FusionDefinition(_C._FusionDefinition):
             self.definition()
             self._finalize_definition()
 
-        if type(self).multidevice_schedule != FusionDefinition.multidevice_schedule:
-            self.multidevice_schedule()
+        self.multidevice_schedule()
 
         # If schedule is defined by child class and schedule is not defined for
         # inputs, make a schedule.

--- a/tests/python/test_multidevice.py
+++ b/tests/python/test_multidevice.py
@@ -56,7 +56,6 @@ def test_sizes_and_ranks(multidevice_test):
     assert local_rank >= 0 and local_rank < local_size
 
 
-@pytest.mark.skip(reason="#2929")
 @pytest.mark.mpi
 def test_pointwise(multidevice_test):
     num_devices = multidevice_test.size
@@ -78,7 +77,7 @@ def test_pointwise(multidevice_test):
             self.t2 = self.ops.add(self.t1, self.t1)
             self.add_output(self.t2)
 
-        def schedule(self):
+        def multidevice_schedule(self):
             mesh = self.sched._create_device_mesh((0, 1))
             self.sched._set_device_mesh(self.t0, mesh)
             self.sched._set_device_mesh(self.t1, mesh)

